### ISSUE TITLE
help request: Config as data_plane problem

### ISSUE THREAD

**Issue by PhanPirang** [apache/apisix#8067]
### Description

I have configured APISIX as decouple mode. The control plane works fine. I can call Admin API however in the data plane container it shows as below:

/usr/local/openresty/luajit/bin/luajit ./apisix/cli/apisix.lua init
failed to validate config: property "etcd" validation failed: property "prefix" is required

Furthermore, where can I find more documents or sample repository relate to deployment mode? I can only find it in the APISIX document.

Here's my config file:

version: 3.0.0-beta

#control_plane config:
```
apisix:
  node_listen: 9080
  enable_ipv6: false

  allow_admin:
    - 0.0.0.0/0

  control:
    ip: "0.0.0.0"
    port: 9092

deployment:
  role: control_plane
  role_control_plane:
    config_provider: etcd
    conf_server:
      listen: 0.0.0.0:9280
      cert: /certs/cp/apisix-cp.pem
      cert_key: /certs/cp/apisix-cp-key.pem

  admin:
    allow_admin:
      - 0.0.0.0/0

  etcd:
    host:                           # it's possible to define multiple etcd hosts addresses of the same etcd cluster.
      - "http://etcd:2379"          # multiple etcd address
    prefix: "/apisix"               # apisix configurations prefix
    timeout: 30                     # 30 seconds
```
    
#data_plane config:

```
apisix:
  node_listen: 9080
  enable_ipv6: false

  enable_control: true
  control:
    ip: "0.0.0.0"
    port: 9092

deployment:
  role: data_plane
  role_data_plane:
    config_provider: control_plane
    control_plane:
      host:
        - "https://apisix-control-plane:9280"
      timeout: 30
  certs:
    cert: /certs/dp/apisix-dp.pem
    cert_key: /certs/dp/apisix-dp-key.pem
    trusted_ca_cert: /certs/ca/rootCA.pem

  admin:
    allow_admin:
      - 0.0.0.0/0

plugin_attr:
  prometheus:
    export_addr:
      ip: "0.0.0.0"
      port: 9091
      
```

#docker compose file:

```
version: "3"
services:
  apisix-dashboard:
    image: apache/apisix-dashboard:2.13-alpine
    restart: always
    volumes:
    - ./dashboard_conf/conf.yaml:/usr/local/apisix-dashboard/conf/conf.yaml
    ports:
    - "9000:9000"
    networks:
      apisix:

  apisix:
    image: apache/apisix:2.99.0-debian
    restart: always
    volumes:
      - ./apisix_dp_log:/usr/local/apisix/logs
      - ./apisix_dp_conf/config.yaml:/usr/local/apisix/conf/config.yaml:ro
      - ./certs/dp:/certs/dp
      - ./certs/ca:/certs/ca
    depends_on:
      - apisix-control-plane
    ##network_mode: host
    ports:
      - "9080:9080/tcp"
      - "9091:9091/tcp"
      - "9443:9443/tcp"
      - "9092:9092/tcp"
    networks:
      apisix:

  apisix-control-plane:
    image: apache/apisix:2.99.0-debian
    restart: always
    volumes:
      - ./apisix_cp_log:/usr/local/apisix/logs
      - ./apisix_cp_conf/config.yaml:/usr/local/apisix/conf/config.yaml:ro
      - ./certs/cp:/certs/cp
      - ./certs/ca:/certs/ca
    depends_on:
      - etcd
    ##network_mode: host
    ports:
      - "9180:9180/tcp"
      - "9280:9280/tcp"
    networks:
      apisix:

  etcd:
    image: bitnami/etcd:3.4.15
    restart: always
    volumes:
      - etcd_data:/bitnami/etcd
    environment:
      ETCD_ENABLE_V2: "true"
      ALLOW_NONE_AUTHENTICATION: "yes"
      ETCD_ADVERTISE_CLIENT_URLS: "http://0.0.0.0:2379"
      ETCD_LISTEN_CLIENT_URLS: "http://0.0.0.0:2379"
    ports:
      - "2379:2379/tcp"
    networks:
      apisix:

  prometheus:
    image: prom/prometheus:v2.25.0
    restart: always
    volumes:
      - ./prometheus_conf/prometheus.yml:/etc/prometheus/prometheus.yml
    ports:
      - "9090:9090"
    networks:
      apisix:

  grafana:
    image: grafana/grafana:7.3.7
    restart: always
    ports:
      - "3000:3000"
    volumes:
      - "./grafana_conf/provisioning:/etc/grafana/provisioning"
      - "./grafana_conf/dashboards:/var/lib/grafana/dashboards"
      - "./grafana_conf/config/grafana.ini:/etc/grafana/grafana.ini"
    networks:
      apisix:

networks:
  apisix:
    driver: bridge

volumes:
  etcd_data:
    driver: local
```

Note: on APISIX 2.15.0, I also got the same problem.

### Environment

- APISIX version (3.0.0-Beta):
- Run as Docker Container

**Comment by spacewander** [comment]
The doc about data plane misses the `prefix` field. It should be:
```
  role: data_plane
  role_data_plane:
    config_provider: control_plane
    control_plane:
      host:
        - "https://apisix-control-plane:9280"
      prefix: "/apisix"
      timeout: 30
```

**Comment by spacewander** [comment]
As another thought, maybe we can make the prefix optional and set it to `/apisix` by default.
What's your opinion? @tzssangglass @leslie-tsang @hf400159

**Comment by tzssangglass** [comment]
> As another thought, maybe we can make the prefix optional and set it to `/apisix` by default.

+1

**Comment by hf400159** [comment]
> As another thought, maybe we can make the prefix optional and set it to `/apisix` by default. 
+1 
And I have fixed this bug. https://github.com/apache/apisix/pull/8072

**Comment by PhanPirang** [comment]
After I added prefix, it works fine, However, there's another problem as below:

request etcd endpoint 'https://apisix-control-plane:9280/version' error, certificate verify failed

Note: I use self-sign certificate from the same Certificate Authority for both data plane and control plane.

**Comment by tzssangglass** [comment]
> After I added prefix, it works fine, However, there's another problem as below:
> 
> request etcd endpoint '[apisix-control-plane:9280/version](https://apisix-control-plane:9280/version)' error, certificate verify failed
> 
> Do I miss any fields?
> 
> Note: I use self-sign certificate from the same Certificate Authority for both data plane and control plane.

which `trusted_ca_cert` is the CA cert that issue your self-sign certificate?

**Comment by tzssangglass** [comment]
read more: https://github.com/apache/apisix/blob/master/docs/en/latest/certificate.md#set-up-multiple-ca-certificates

**Comment by PhanPirang** [comment]
> > After I added prefix, it works fine, However, there's another problem as below:
> > request etcd endpoint '[apisix-control-plane:9280/version](https://apisix-control-plane:9280/version)' error, certificate verify failed
> > Do I miss any fields?
> > Note: I use self-sign certificate from the same Certificate Authority for both data plane and control plane.
> 
> which `trusted_ca_cert` is the CA cert that issue your self-sign certificate?

Yes, `trusted_ca_cert` is the root CA cert that signed all the certificates.

**Comment by tzssangglass** [comment]
If there is no need for confidentiality, you can post the certificate

**Comment by PhanPirang** [comment]
> If there is no need for confidentiality, you can post the certificate

Which certificate that you want to see?

**Comment by tzssangglass** [comment]
> request etcd endpoint '[apisix-control-plane:9280/version](https://apisix-control-plane:9280/version)' error, certificate verify failed

All certs associated with this error so that I can verify with the same certificate.

**Comment by PhanPirang** [comment]
[mkcert.zip](https://github.com/apache/apisix/files/9761131/mkcert.zip)

Here I use mkcert.

**Comment by PhanPirang** [comment]
> > request etcd endpoint '[apisix-control-plane:9280/version](https://apisix-control-plane:9280/version)' error, certificate verify failed
> 
> All certs associated with this error so that I can verify with the same certificate.

[mkcert.zip](https://github.com/apache/apisix/files/9761136/mkcert.zip)

I use mkcert.

**Comment by PhanPirang** [comment]
Could you share some samples on setting up Decouple Mode?

**Comment by tzssangglass** [comment]
> Decouple Mode

What do you mean by Decouple Mode?

**Comment by PhanPirang** [comment]
I mean sample configuration for setting up the data plane and control plane separately.

**Comment by PhanPirang** [comment]
> > Decouple Mode
> 
> What do you mean by Decouple Mode?

I mean sample configuration for setting up the data plane and control plane separately.

**Comment by PhanPirang** [comment]
> request etcd endpoint '[apisix-control-plane:9280/version](https://apisix-control-plane:9280/version)' error, certificate verify failed

Do you have any advice on this problem?

**Comment by tzssangglass** [comment]
> Do you have any advice on this problem?

The `apisix-cp.pem` you generated does not have a Common Name. 

```yaml
    control_plane:
      host:
        - "https://apisix-control-plane:9280"
```

APISIX would verify the `apisix-control-plane` is the same as the Common Name of `apisix-cp.pem`.

you can disable this verification by add `tls.verify`

```
    control_plane:
      host:
        - "https://apisix-control-plane:9280"
      timeout: 30
      prefix: "/apisix"
      tls:
        verify: false
```

Or you need to regenerate the certificate that issued the Common Name.

**Comment by tzssangglass** [comment]
my data_plane config:

```yaml
apisix:
  ssl:
    ssl_trusted_certificate: t/certs/mtls_ca.crt
deployment:
  role: data_plane
  role_data_plane:
    config_provider: control_plane
    control_plane:
      host:
        - "https://admin.apisix.dev:9280"
      timeout: 30
      prefix: "/apisix"
```

NOTE: 

```
apisix:
  ssl:
    ssl_trusted_certificate: t/certs/mtls_ca.crt
```
ssl_trusted_certificate set the CA cert that issued the cert of conf_server in the control plane.


my control_plane config:

```
deployment:
  role: control_plane
  role_control_plane:
    config_provider: etcd
    conf_server:
      listen: 0.0.0.0:9280
      cert: t/certs/mtls_server.crt
      cert_key: t/certs/mtls_server.key
  admin:
    allow_admin:
      - 0.0.0.0/0
    admin_key:
      - name: admin
        key: edd1c9f034335f136f87ad84b625c8f1  # using fixed API token has security risk, please update it when you deploy to production environment
        role: admin
```

cat /etc/hosts:

```
10.211.55.4 admin.apisix.dev
```

**Comment by PhanPirang** [comment]
> 

Thanks, let me try!

**Comment by PhanPirang** [comment]
@tzssangglass Hello, I have tried, but I still got the same error "certificate verify failed". Can you share a sample repository?

**Comment by tzssangglass** [comment]
> @tzssangglass Hello, I have tried, but I still got the same error "certificate verify failed". Can you share a sample repository?

Did you use your own certificate or `t/certs/mtls_server.crt` when you tried?

**Comment by PhanPirang** [comment]
> > @tzssangglass Hello, I have tried, but I still got the same error "certificate verify failed". Can you share a sample repository?
> 
> Did you use your own certificate or `t/certs/mtls_server.crt` when you tried?

I used my own certificate.

**Comment by tzssangglass** [comment]
> I used my own certificate.

try https://github.com/apache/apisix/blob/master/t/certs/mtls_server.crt as my demo.

if you use your own certificate, you need to change the host to your certs' common name.

**Comment by jinjianming** [comment]
@tzssangglass 
hello tzssangglass :
  I encountered the same problem when deploying in K8S mode;
Even if my certificate CN information is the same as the apisix-control-plane name, it cannot be authenticated successfully。
I finally added ` tls.verify: false`
```
https://github.com/apache/apisix-helm-chart/issues/479
```

**Comment by jinjianming** [comment]
> @tzssangglass您好，我已经尝试过了，但仍然出现“证书验证失败”的相同错误。你能分享一个示例存储库吗？

You need to set the_ Plane configuration adding client certificate
```
  certs:
    cert: /certs/dp/apisix-dp.pem
    cert_key: /certs/dp/apisix-dp-key.pem
```

**Comment by XW512** [comment]
> my data_plane config:
> 
> ```yaml
> apisix:
>   ssl:
>     ssl_trusted_certificate: t/certs/mtls_ca.crt
> deployment:
>   role: data_plane
>   role_data_plane:
>     config_provider: control_plane
>     control_plane:
>       host:
>         - "https://admin.apisix.dev:9280"
>       timeout: 30
>       prefix: "/apisix"
> ```
> 
> NOTE:
> 
> ```
> apisix:
>   ssl:
>     ssl_trusted_certificate: t/certs/mtls_ca.crt
> ```
> 
> ssl_trusted_certificate set the CA cert that issued the cert of conf_server in the control plane.
> 
> my control_plane config:
> 
> ```yaml
> deployment:
>   role: control_plane
>   role_control_plane:
>     config_provider: etcd
>     conf_server:
>       listen: 0.0.0.0:9280
>       cert: t/certs/mtls_server.crt
>       cert_key: t/certs/mtls_server.key
>   admin:
>     allow_admin:
>       - 0.0.0.0/0
>     admin_key:
>       - name: admin
>         key: edd1c9f034335f136f87ad84b625c8f1  # using fixed API token has security risk, please update it when you deploy to production environment
>         role: admin
> ```
> 
> cat /etc/hosts:
> 
> ```
> 10.211.55.4 admin.apisix.dev
> ```
> 
> `10.211.55.4` is the IP of control plane, and `admin.apisix.dev` is the Common Name of `mtls_server.crt`.

Hello, according to your configuration, I found that the data of the CP node has been modified, but the DP node did not automatically pull the latest configuration. If the configuration data is the latest after restarting the DP node, is it not quasi real-time to synchronize to the DP node?

**Comment by XW512** [comment]
> > > @tzssangglass Hello, I have tried, but I still got the same error "certificate verify failed". Can you share a sample repository?
> > 
> > 
> > Did you use your own certificate or `t/certs/mtls_server.crt` when you tried?
> 
> I used my own certificate.

Have you solved this problem? I think there should be no problem with MTLS communication on my end, but the CP node has updated data and the DP node cannot automatically refresh. Have you encountered this problem?

**Comment by kingluo** [comment]
The conf server is removed, so the etcd communication rolls back to what 2.x versions behave like, i.e. no matter CP or DP, connects to etcd directly. Closed by #10012

**Comment by kingluo** [comment]
The conf server is removed, so the etcd communication rolls back to what 2.x versions behave like, i.e. no matter CP or DP, connects to etcd directly. Closed by https://github.com/apache/apisix/pull/10012

**Comment by jonforero** [comment]
> The conf server is removed, so the etcd communication rolls back to what 2.x versions behave like, i.e. no matter CP or DP, connects to etcd directly. Closed by #10012

From version 3, the communication between Data planes and Control Plane is no more supported? I mean, Now the communication of the data planes need to communicate directly to etcd?

**Comment by jonforero** [comment]
I received this message:

2024-07-31 12:38:14 invalid deployment data_plane configuration: property "role_data_plane" validation failed: property "config_provider" validation failed: matches none of the enum values

deployment:
  role: data_plane
  role_data_plane:
    config_provider: control_plane
    control_plane:
      host:
        - "http://apisix-cp:9280"
      prefix: "/apisix"
      timeout: 30
      tls:
        verify: false
    certs: {}
